src/main/client/scene/game-scene.ts

```diff
@@ -99,30 +83,120 @@ export class GameScene extends Scene {
          alpha: canvasData.data[index + 3],
       };
    }
+   private createHole(worldX: number, worldY: number): void {
+      this.terrainTexture.context.globalCompositeOperation = 'destination-out';
+      this.terrainTexture.context.beginPath();
+      this.terrainTexture.context.arc(worldX, worldY, 30, 0, Math.PI * 2, true);
+      this.terrainTexture.context.fill();
+
+      const newCanvasData = this.terrainTexture.context.getImageData(
+         0,
+         0,
+         this.terrainTexture.getSourceImage().width,
+         this.terrainTexture.getSourceImage().height,
+      );
+
+      this.terrainTexture.clear();
+
+      this.terrainTexture.imageData = newCanvasData;
+      this.terrainTexture.putData(newCanvasData, 0, 0);
+      this.terrainTexture.refresh();
+   }
 
    private jumping = false;
    private verticalSpeed = 0;
 
+   private getDownwardVector(sprite: Phaser.GameObjects.Sprite): Phaser.Math.Vector2 {
+      return new Phaser.Math.Vector2({ x: -sprite.x, y: -sprite.y }).normalize();
+   }
+
+   private getFloorVector(sprite: Phaser.GameObjects.Sprite): Phaser.Math.Vector2 {
+      return this.getDownwardVector(sprite).normalizeRightHand();
+   }
+
+   private applyGravity(sprite: Phaser.GameObjects.Sprite): void {
+      const vector = this.getDownwardVector(sprite).scale(-1);
+      this.moveByVector(sprite, vector);
+   }
+
+   private applyGroundReactionForce(sprite: Phaser.GameObjects.Sprite): void {
+      const vector = this.getDownwardVector(sprite).scale(0.5);
+      this.moveByVector(sprite, vector);
+   }
+
+   private moveByVector(sprite: Phaser.GameObjects.Sprite, vector: Phaser.Math.Vector2): void {
+      sprite.x += vector.x;
+      sprite.y += vector.y;
+   }
+
+   private createLocalWall(sprite: Phaser.GameObjects.Sprite, length: number, localOffset: Phaser.Geom.Point): Phaser.Geom.Point[] {
+      const downVector = this.getDownwardVector(sprite);
+
+      return this.createCollisionLine(downVector, length, -length);
+   }
+
+   private createLocalFloor(sprite: Phaser.GameObjects.Sprite, length: number, localOffset: Phaser.Geom.Point): Phaser.Geom.Point[] {
+      const floorVector = this.getFloorVector(sprite);
+
+      return this.createCollisionLine(floorVector, length, -length / 2);
+   }
+
+   private createCollisionLine(vector: Phaser.Math.Vector2, length: number, offset: number): Phaser.Geom.Point[] {
+      return [...Array(length).keys()]
+         .map((key) => key + offset)
+         .map((i) => {
+            const newDownVector = vector.clone().scale(i);
+            return new Phaser.Geom.Point(newDownVector.x, newDownVector.y);
+         });
```

Comment:
You dont need to create an extra array from keys(), it is already a new array.
I think one map is enough, for example (key)=>{const i=key+offset...

Reply:
yeah one map probs enough, but i found this more readable, i could change it to one map if thats easier for u